skyportal/models.py

```diff
@@ -1645,39 +1652,170 @@ class Spectrum(Base):
-            raise ValueError(f"Expected 2 columns, got {data.shape[1]}")
+
+        filename = Path(filename)
+
+        if data is None:
+            try:
+                table = ascii.read(filename, comment='#', header_start=None)
+            except Exception as e:
+                e.args = (f'Error parsing ASCII file: {e.args[0]}',)
+                raise
+            bytes = open(filename).read().encode('ascii')
+        else:
+            file_obj = io.BytesIO(data.encode('ascii'))
+            try:
+                table = ascii.read(file_obj, comment='#', header_start=None)
+            except Exception as e:
+                e.args = (f'Error parsing ASCII file: {e.args[0]}',)
+                raise
+            bytes = data
+
+        tabledata = np.asarray(table)
+        colnames = table.colnames
+
+        if len(colnames) < 2:
+            raise ValueError(
+                'Input data must have at least 2 columns (wavelength, '
+                'flux, and optionally flux error).'
+            )
+
+        if 'comments' in table.meta:
+
+            # this section matches lines like:
+            # XTENSION: IMAGE
+            # BITPIX: -32
+            # NAXIS: 2
+            # NAXIS1: 433
+            # NAXIS2: 1
+
+            header = {}
+            for line in table.meta['comments']:
+                try:
+                    result = yaml.load(line, Loader=yaml.FullLoader)
+                except yaml.YAMLError:
+                    continue
+                if isinstance(result, dict):
+                    header.update(result)
+
+            # otherwise read it like a fits header
+            cards = []
+            with warnings.catch_warnings():
+                warnings.simplefilter('error', AstropyWarning)
+                for line in table.meta['comments']:
+                    # this line does not raise a warning
+                    card = fits.Card.fromstring(line)
+                    try:
+                        # this line warns (exception in this context)
+                        card.verify()
+                    except AstropyWarning:
+                        continue
+                    cards.append(card)
+
+            # this ensures lines like COMMENT and HISTORY are properly dealt with
+            fits_header = fits.Header(cards=cards)
+            serialized = dict(fits_header)
+
+            for key in serialized:
+                if isinstance(serialized[key], fits.header._CardAccessor):
+                    # serialize as a string
+                    serialized[key] = str(serialized[key])
+                if len(fits_header.comments[key]) > 0:
+                    header[key] = {
+                        'value': serialized[key],
+                        'comment': fits_header.comments[key],
+                    }
+                else:
+                    header[key] = serialized[key]
+
+            # coerce things to serializable JSON
```

Comment:
Our JSON writer can already handle this?

Reply:
I remember having a good reason for putting this in when I did, but I commented it out and the tests pass.

Reply:
Now I remember why I put this in. It's for serializing *to* the database, not for serializing *from* the database. I.e., to avoid errors like ```JSON response: {‘data’: {}, ‘version’: ‘0.9.dev0’, ‘status’: ‘error’, ‘message’: “(builtins.TypeError) Object of type datetime is not JSON serializable\n[SQL: INSERT INTO spectra (created_at, modified, wavelengths, fluxes, errors, obj_id, observed_at, origin, instrument_id, followup_request_id, assignment_id, altdata, original_file_string, original_file_filename) VALUES (timezone(%(timezone_1)s, CURRENT_TIMESTAMP), timezone(%(timezone_1)s, CURRENT_TIMESTAMP), %(wavelengths)s, %(fluxes)s, %(errors)s, %(obj_id)s, %(observed_at)s, %(origin)s, %(instrument_id)s, %(followup_request_id)s, %(assignment_id)s, %(altdata)s, %(original_file_string)s, %(original_file_filename)s) RETURNING spectra.id]\n[parameters: [{‘original_file_string’: ‘# SIMPLE: True\\n# BITPIX: -64\\n# NAXIS: 2\\n# NAXIS1: 214\\n# NAXIS2: 220\\n# EXTEND: True\\n# CDELT1: 25.570776255704914\\n# CRVAL1: 3776.7123287671147\\n ... (9831 characters truncated) ... 556e-33\\n9121.0 1.591e-16 1.680e-33\\n9146.6 1.619e-16 1.214e-33\\n9172.1 1.528e-16 1.309e-33\\n9197.7 1.428e-16 1.388e-33\\n9223.3 1.357e-16 1.630e-33\\n’, ‘observed_at’: datetime.datetime(2020, 9, 1, 5, 42, 29, tzinfo=datetime.timezone.utc), ‘obj_id’: ‘ZTF20abxgwjr’, ‘altdata’: {‘SIMPLE’: True, ‘BITPIX’: -64, ‘NAXIS’: 2, ‘NAXIS1’: 214, ‘NAXIS2’: 220, ‘EXTEND’: True, ‘CDELT1’: 25.570776255704914, ‘CRVAL1’: 3776.7123287671147,  ... (3960 characters truncated) ...  ‘SNIDAGEMED’: -6.15, ‘SNIDAGEMEDERR’: 5.38, ‘SNIDFRAC_IA’: 0.97, ‘SNIDFRAC_IB’: 0.01, ‘SNIDFRAC_IC’: 0.02, ‘SNIDFRAC_II’: 0.0, ‘SNIDFRAC_NOTSN’: 0.0}, ‘instrument_id’: 2, ‘wavelengths’: array([3776.7, 3802.3, 3827.9, 3853.4, 3879. , 3904.6, 3930.1, 3955.7,\n       3981.3, 4006.8, 4032.4, 4058. , 4083.6, 4109.1, 4134.7, 4160.3,\n       4 ... (1601 characters truncated) ... 8814.2, 8839.7, 8865.3,\n       8890.9, 8916.4, 8942. , 8967.6, 8993.2, 9018.7, 9044.3, 9069.9,\n       9095.4, 9121. , 9146.6, 9172.1, 9197.7, 9223.3]), ‘fluxes’: array([ 5.403e-16,  1.065e-16, -6.951e-17,  1.738e-16,  3.601e-16,\n        3.258e-16,  3.241e-16,  4.001e-16,  4.439e-16,  5.263e-16,\n        7.520e-1 ... (2569 characters truncated) ... 16,  1.504e-16,  1.566e-16,\n        1.523e-16,  1.276e-16,  1.227e-16,  1.423e-16,  1.591e-16,\n        1.619e-16,  1.528e-16,  1.428e-16,  1.357e-16]), ‘original_file_filename’: ‘spec_auto_contsep_lstep1__crr_b_ifu20200901_05_42_29_ZTF20abxgwjr.txt’, ‘errors’: None, ‘assignment_id’: None, ‘followup_request_id’: None, ‘origin’: None}]]“}
```